### PR TITLE
Migrate 'International > Locations > States' Listing

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Migrate 'International > Locations > States' Listing
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Relative to #10606
| How to test?      | The listing is now available in the URL `/admin-dev/index.php/improve/international/states/` :<br>* Pagination<br>* Enable/Disable (Bulk or not)<br>* Delete (Bulk or not)<br>* Search<br>* Add & Edit (always in legacy)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27825)

### CODE REVIEW

File: src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+class ZoneByIdChoiceProvider implements ConfigurableFormChoiceProviderInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getChoices(array $options): array
+    {
+        $active = isset($options['active']) && is_bool($options['active']) && $options['active'];
+
+        $activeFirst = isset($options['active_first']) && is_bool($options['active_first']) && $options['active_first'];
```
Comment (saulaski) [inline src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php]: If bool type is expected as option, an exception shouldn't get thrown in case of another type provided ?
```suggestion
        if (isset($options['active']) && !is_bool($options['active']) {
            new InvalidArgumentException(
                sprintf('"active" option type should be "bool", got "%s" type instead', gettype($options['active']));
        }
        $active = $options['active'] ?? false;

        if (isset($options['activeFirst']) && !is_bool($options['activeFirst']) {
            new InvalidArgumentException(
                sprintf('"activeFirst" option type should be "bool", got "%s" type instead', gettype($options['activeFirst']));
        }
        $activeFirst = $options['activeFirst'] ?? false;
```

File: src/Core/Form/ChoiceProvider/CountryByIdConfigurableChoiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * {@inheritdoc}
+     */
+    public function getChoices(array $options): array
+    {
+        $active = isset($options['active']) && is_bool($options['active']) && $options['active'];
+
+        $containsStates = isset($options['contains_states']) && is_bool($options['contains_states']) && $options['contains_states'];
+
+        $listStates = isset($options['list_states']) && is_bool($options['list_states']) && $options['list_states'];
```
Comment (saulaski) [inline src/Core/Form/ChoiceProvider/CountryByIdConfigurableChoiceProvider.php]: Same question: non boolean values shouldn't throw an exception ?

File: src/Core/Grid/Query/StateQueryBuilder.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * Apply filters to state query builder.
+     *
+     * @param array $filters
+     * @param QueryBuilder $qb
+     */
+    private function applyFilters(QueryBuilder $qb, array $filters): void
+    {
+        dump($qb);
+        dump($filters);
```
Comment (saulaski) [inline src/Core/Grid/Query/StateQueryBuilder.php]: oops :)

File: src/Core/Grid/Query/StateQueryBuilder.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+        $qb
+            ->select('s.`id_state`, s.`name`, s.`iso_code`, s.`id_country`, s.`id_zone`, s.`active`')
+            ->addSelect('cl.`name` as country_name, z.`name` as zone_name')
```
Comment (saulaski) [inline src/Core/Grid/Query/StateQueryBuilder.php]: Or use only 1 method with 1 argument / line

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 5]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @param Request $request
+     * @param StateFilters $filters
+     *
+     * @return Response
+     */
+    public function indexAction(Request $request, StateFilters $filters): Response
+    {
+        dump($filters);
```
Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: oops :)

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 6]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     * @return Response
+     */
+    public function editAction(int $stateId, Request $request): Response
+    {
+        // todo: replace when https://github.com/PrestaShop/PrestaShop/pull/15071 is merged
+        return $this->redirect(
+            $this->getAdminLink(
+                'AdminStates',
+                ['updatestate' => '', 'id_state' => $stateId],
+                true)
```
Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
            $this->getAdminLink(
                'AdminStates',
                ['updatestate' => '', 'id_state' => $stateId],
                true
            )
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 7]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+
+        return $stateIds;
+    }
+
+    /**
+     * @param Exception|null $e
+     *
+     * @return array
+     */
+    private function getErrorMessages(Exception $e = null): array
```
Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: Just in case of exotic exception
```suggestion
    private function getErrorMessages(Throwable $e = null): array
```

Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: What's the point of the argument that isn't used in the method ?

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml [hunk 8]
```diff
@@ -1,6 +1,90 @@
...
+    _legacy_controller: AdminStates
+
+admin_states_index:
   path: /
   methods: [ GET ]
   defaults:
-    _controller: 'PrestaShopBundle\Controller\Admin\Improve\International\StateController::getStatesAction'
+    _controller: PrestaShopBundle:Admin\Improve\International\State:index
+    _legacy_controller: AdminStates
+#    _legacy_link: AdminStates
```
Comment (saulaski) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml:14]: Is this commented line to be kept ? same about other commented lines

**Review by saulaski — CHANGES_REQUESTED** [review]
LGTM overall, only minor issues.

File: src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php (hunk 1 above)
Comment (jolelievre) [reply src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php]: Not mandatory I'd say because the providers are mostly used in a certain context, in the core forms that respect the expected input options, but it's a good suggestion nonetheless (even if there are few chances for it to be used somewhere else or by anyone else)

Comment (jolelievre) [reply src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php]: You could also use an `OptionsResolver` component to handle these input options

File: src/Core/Form/ChoiceProvider/CountryByIdConfigurableChoiceProvider.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Form/ChoiceProvider/CountryByIdConfigurableChoiceProvider.php]: Same suggestion here, maybe use an `OptionsResolver`

File: src/Core/Grid/Query/StateQueryBuilder.php (hunk 4 above)
Comment (jolelievre) [reply src/Core/Grid/Query/StateQueryBuilder.php]: I guess this was done to avoid using too many columns for readability, I think it's acceptable

File: src/Core/Grid/Query/StateQueryBuilder.php (hunk 3 above)
Comment (jolelievre) [reply src/Core/Grid/Query/StateQueryBuilder.php]: Hehe, busted!! 😅

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 9]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     *     message="You do not have permission to create this."
+     * )
+     *
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function createAction(Request $request): Response
+    {
+        // todo: replace when https://github.com/PrestaShop/PrestaShop/pull/15071 is merged
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: Hmm, be careful https://github.com/PrestaShop/PrestaShop/pull/15071 was closed months ago..

I don't know if it should be reopened to be finished I don't even understand why it's been closed actually the reason comes from another PR which doesn't seem related 🤷‍♂️

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 10]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     *
+     * @param int $stateId
+     *
+     * @return RedirectResponse
+     */
+    public function toggleStatusAction(int $stateId): RedirectResponse
+    {
+        try {
+            /** @var EditableState $editableState */
+            $editableState = $this->getQueryBus()->handle(new GetStateForEditing($stateId));
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: `$editableState` is not used since `ToggleStateStatusCommand` doesn't seem to care about the current state It simply toggles it 🤷‍♂️

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 11]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     * Toggles state status
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", redirectRoute="admin_states_index")
+     * @DemoRestricted(redirectRoute="admin_states_index")
+     *
+     * @param int $stateId
+     *
+     * @return RedirectResponse
+     */
+    public function toggleStatusAction(int $stateId): RedirectResponse
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: Just a suggestion for the "in row toggle" Maybe you could do the same thing as @matks did here https://github.com/PrestaShop/PrestaShop/pull/25115

Basically the idea is to not refresh the whole page for a single toggle but to perform an ajax toggle, from what I understood it means:
- using the `AsyncToggleColumnExtension` js component instead of the `ColumnTogglingExtension` one
- this controller must be constructed as an API and return a JSON response

It's a nice improvement which has only been introduced after the initial `ColumnTogglingExtension` but I'd say it's better from a UX perspective

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: That's because in most of our controller the `getErrorMessages` does use the exception for some of the error messages I guess this one was copied from another controller

Although in this case the parameter is not used so it could either be removed for now, or kept as an anticipation of the future error messages to integrate

That said very few error messages have been handled here, you need to check for each controller (and behind it for each CQRS command) which exception can be thrown an handle an appropriate error message

For example `DeleteStateException` can be thrown (by both regular delete AND bulk delete) and it contains the state ID which could be included in the error messages

Although from what I see in the bulk handlers the exception are not ideally handled, they should not stop at the first error but instead save the failed IDs keep doing their action and finally throw an exception with the list of failed IDs which can then be displayed in the error message But it's not the purpose of this PR to modify this, especially since we lack some convention and consistency in how we handle bulk commands (sometimes it's "fail at first" sometimes it's "fail at the end")

Regardless you can handle more messages in this method, and once they are all handled if the exception is still not used then it can be removed as a parameter (we can always reintroduce it later if it's one day needed)

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml [hunk 12]
```diff
@@ -1,6 +1,90 @@
 admin_country_states:
+  path: /country-states
+  methods: [ GET ]
+  defaults:
+    _controller: 'PrestaShopBundle:Admin\Improve\International\State:getStates'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]:
```suggestion
    _controller: PrestaShopBundle\Controller\Admin\Improve\International\StateController::getStatesAction
```

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]: Please use this syntax for routing everywhere, the former one is deprecated and will be removed

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml (hunk 8 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml:14]: For `_legacy_link` we keep them commented at first it allows to access the new page (only via its URL) and the legacy controller as well It's useful for QA to compare both pages

The `_legacy_link` is usually enabled when the whole page (list + form) has been completely migrated Often it's done in a last dedicated PR which removes legacy code and enabled the links

File: src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml [hunk 13]
```diff
@@ -409,3 +409,11 @@ services:
...
       - "@=service('prestashop.adapter.legacy.context').getContext().language.id"
     public: true
+
+  prestashop.core.grid.query_builder.state:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Query\StateQueryBuilder'
+    parent: 'prestashop.core.grid.abstract_query_builder'
+    public: true
+    arguments:
+      - '@prestashop.core.query.doctrine_search_criteria_applicator'
+      - "@=service('prestashop.adapter.legacy.context').getContext().employee.id_lang"
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:419]: Curious question but why not use `@=service("prestashop.adapter.legacy.context").getLanguage().id` here?

Relying on employee field can be a bit unstable in test environment

**Review by jolelievre — COMMENTED** [review]
Thanks @Progi1984 

Nice migration, a few code to clean and some suggestions for improvement

File: src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php (hunk 1 above)
Comment (saulaski) [reply src/Adapter/Form/ChoiceProvider/ZoneByIdChoiceProvider.php]: `OptionResolver` is an excellent idea. Most helpful if it was included in a new `AbstractConfigurableChoiceProvider` class to extend, as a template method pattern, just like Symfony form types.

File: src/Core/Grid/Query/StateQueryBuilder.php (hunk 3 above)
Comment (saulaski) [reply src/Core/Grid/Query/StateQueryBuilder.php]: Note for myself, or for anybody who got free time: include `dump()` amongst forbidden keywords in CI tests :-)

File: src/Core/Grid/Query/StateQueryBuilder.php (hunk 4 above)
Comment (saulaski) [reply src/Core/Grid/Query/StateQueryBuilder.php]: 1 argument per line is usually the best option to limit line length and improve readibility.
Here an extra method call impacts performance only for code style reasons.

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php (hunk 7 above)
Comment (saulaski) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: @jolelievre Understood. Then the argument here is useless for now.

About the bulk delete behaviour, there is no absolute rule. Only a question of functional choice:
1. do nothing if a single error happens
2. do as many as possible, as you suggest
3. do until first error happens, then stop

UX/UI should be taken in consideration in such choice. What is the behavior the end user needs the most ?
Usually the most common choice is all or nothing (option 1) to help the end user who then doesn't have to search what's left to delete. Flash notification disappears too quickly, and the user has no time to take note.

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml (hunk 8 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml:14]: :+1:

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php (hunk 11 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]: :+1:

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml (hunk 12 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]: :+1:

File: src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml (hunk 13 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:419]: With the `employee.id_lang"`, you will have the list of countries in the language of the employee.

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 14]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     *
+     * @return RedirectResponse
+     */
+    public function deleteAction(int $stateId): RedirectResponse
+    {
+        try {
+            $this->getCommandBus()->handle(new DeleteStateCommand($stateId));
+            $this->addFlash(
+                'success',
+                $this->trans('Successful deletion.', 'Admin.Notifications.Success')
```
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
                $this->trans('Successful deletion', 'Admin.Notifications.Success')
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 15]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+     */
+    public function deleteBulkAction(Request $request): RedirectResponse
+    {
+        $stateIds = $this->getBulkStatesFromRequest($request);
+
+        try {
+            $this->getCommandBus()->handle(new BulkDeleteStateCommand($stateIds));
+            $this->addFlash(
+                'success',
+                $this->trans('Successful deletion.', 'Admin.Notifications.Success')
```
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
                $this->trans('Successful deletion', 'Admin.Notifications.Success')
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 16]
```diff
@@ -64,4 +76,266 @@ public function getStatesAction(Request $request)
...
+    /**
+     * @param Exception|null $e
+     *
+     * @return array
+     */
+    private function getErrorMessages(Exception $e = null): array
+    {
+        return [
+            StateNotFoundException::class => $this->trans(
+                'The object cannot be loaded (or found)',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
                'The object cannot be loaded (or found).',
```

**Review by Julievrz — COMMENTED** [review]
Small details 😇

**Review by Julievrz — DISMISSED** [review]
Wording ✅

**Review by matks — COMMENTED** [review]
So sanity tests are failing... for a reason ... really ... strange

```
Error: .github/workflows/sanity.yml (Line: 64, Col: 16):
Error: The template is not valid. .github/workflows/sanity.yml (Line: 64, Col: 16): hashFiles('**/composer.lock') failed. Fail to hash files under directory '/home/runner/work/PrestaShop/PrestaShop'
```

Sanity file -> https://github.com/PrestaShop/PrestaShop/blob/develop/.github/workflows/sanity.yml

I cannot understand the relationship between your code and the error. I restart the CI.

**Comment by saulaski** [comment]
@matks CI issue is not related to this only PR, but to a few PRs in progress

**Review by matthieu-rolland — DISMISSED** [review]

**Review by saulaski — DISMISSED** [review]

**Review by tegbessou — DISMISSED** [review]
Very nice work @Progi1984!!!

**Comment by Progi1984** [comment]
@tegbessou @saulaski @matthieu-rolland : Rebased for fixing the CI

**Review by matthieu-rolland — DISMISSED** [review]

**Review by saulaski — DISMISSED** [review]

File: src/Core/Grid/Definition/Factory/StateGridDefinitionFactory.php [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+                    ->setAssociatedColumn('zone_name')
+            )
+            ->add((new Filter('id_country', ConfigurableCountryChoiceType::class))
+            ->setTypeOptions([
+                'required' => false,
+                'contains_states' => true,
+                'choice_translation_domain' => false,
+            ])
+            ->setAssociatedColumn('country_name')
+            )
```
Comment (atomiix) [inline src/Core/Grid/Definition/Factory/StateGridDefinitionFactory.php]: I think you can improve the indentation here ;)

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 18]
```diff
@@ -64,4 +75,269 @@ public function getStatesAction(Request $request)
...
+                true
+            )
+        );
+    }
+
+    /**
+     * Show "Add new" form and handle form submit.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['create'], request.get('_legacy_controller'))",
```
Comment (atomiix) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
     *     "is_granted('create', request.get('_legacy_controller'))",
```

File: src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php [hunk 19]
```diff
@@ -64,4 +75,269 @@ public function getStatesAction(Request $request)
...
+
+        return $stateIds;
+    }
+
+    /**
+     * @param Throwable|null $e
+     *
+     * @return array
+     */
+    private function getErrorMessages(Throwable $e = null): array
```
Comment (atomiix) [inline src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php]:
```suggestion
    private function getErrorMessages(?Throwable $e = null): array
```

File: src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        // Set normalizer enables to use closure for choice generation with options
+        $resolver->setNormalizer(
+            'choices', function (Options $options) {
```
Comment (atomiix) [inline src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php:62]:
```suggestion
            'choices', static function (Options $options) {
```

File: src/PrestaShopBundle/Form/Admin/Type/ZoneChoiceType.php [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        // Set normalizer enables to use closure for choice generation with options
+        $resolver->setNormalizer(
+            'choices', function (Options $options) {
```
Comment (atomiix) [inline src/PrestaShopBundle/Form/Admin/Type/ZoneChoiceType.php:62]:
```suggestion
            'choices', static function (Options $options) {
```

**Review by atomiix — DISMISSED** [review]

**Review by saulaski — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php (hunk 20 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php:62]: I agree, although it's not blocking

File: src/PrestaShopBundle/Form/Admin/Type/ZoneChoiceType.php (hunk 21 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Type/ZoneChoiceType.php:62]: SAme, and since there is conflict on the PR anyway 🤷‍♂️

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml [hunk 22]
```diff
@@ -1,6 +1,90 @@
...
+    _legacy_controller: AdminStates
+#    _legacy_link: AdminStates:submitBulkdeletestate
+
+admin_states_delete:
+  path: /{stateId}/delete
+  methods: [ DELETE ]
+  defaults:
+    _controller: PrestaShopBundle\Controller\Admin\Improve\International\StateController::deleteAction
+    _legacy_controller: AdminStates
+    _legacy_link: AdminStates:deletestate
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]:
```suggestion
#    _legacy_link: AdminStates:deletestate
```

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]: Shouldn't this one be commented as well for now?

**Review by jolelievre — COMMENTED** [review]
Thanks @Progi1984 

All good for me, just a few suggestions to be sure They are not mandatory but since the PR has conflict anyway maybe you can find extra time to integrate them

File: src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php (hunk 20 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Type/ConfigurableCountryChoiceType.php:62]: No `static` in this case, because i use a property of the class parent inside the anonymous function.

File: src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml (hunk 22 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Resources/config/routing/admin/improve/international/state.yml]: Fixed :fearful:

**Review by saulaski — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @Progi1984

**Review by matthieu-rolland — APPROVED** [review]

**Review by AureRita — CHANGES_REQUESTED** [review]
Hello @Progi1984, 

I tested it and : 
- Search
- Add & Edit 
- The link is correct

works correctly but : 
- Pagination seems to be a little bit different, it needs more number i think  
- When you check the first checkbox ( to check them all ) that doesn't check them 
- When you check some checkbox, the bulk actions still be unusable
- The Disable option isn't save so I can't test the Enable mode 
- When you try to delete something (single option) , That doesn't delete the states and you're moved on the top of the webpage 

As you can see : 

Name on the link
image

Pagination with only your page and the last page  
image

Bulk action disabled and checkbox 
https://user-images.githubusercontent.com/100699445/157868891-a3a249ab-f05f-4640-a8a8-b3c9dc598cbb.mp4

Enable/Disable ( the search mode works ) 
https://user-images.githubusercontent.com/100699445/157869372-ab7de773-8592-41f3-80a3-108b8b41c151.mp4

Delete action, Add/ Edit action 
https://user-images.githubusercontent.com/100699445/157871020-1f4ffeed-e983-4d15-9210-10ff93eeb421.mp4

**Comment by Progi1984** [comment]
@AureRita

> Pagination seems to be a little bit different, it needs more number i think

Nope, it's the same behavior for each migrated grid. You can check it on Taxes page (`/admin-dev/index.php/improve/international/taxes/`)
image

> When you check the first checkbox ( to check them all ) that doesn't check them
> When you check some checkbox, the bulk actions still be unusable
> The Disable option isn't save so I can't test the Enable mode
>When you try to delete something (single option) , That doesn't delete the states and you're moved on the top of the webpage

Could you check again ? I have build my assets this morning, and I haven't got the problem

**Review by AureRita — APPROVED** [review]
Thank you @Progi1984, 

Indeed, when I tested it with a make assets, it works, 

as you can see on this vidéo: 

All negative point disappear : 
https://user-images.githubusercontent.com/100699445/158139443-fff54d3f-996b-4051-83c3-289890584bd4.mp4

**Comment by Progi1984** [comment]
Thanks @AureRita
